Installer for Windows (NSIS-based)

gianluca-nitti opened the pull request:
**Edit: No longer WIP, I committed on this branch all I had in mind for this feature - waiting for review.**

This is an installer for the Windows OS, based on [NSIS](http://nsis.sourceforge.net/Main_Page).

It allows the user to install the launcher with options to create desktop and Start Menu shortcuts and changing the installation directory. Before installing, it ensures a supported Java installation (JRE or JDK) is available on the system; if it's not, it asks the user to download the latest JRE from the official Java website (some screenshots of the tests I made on a cleanly installed Win10 VM ara available [here](http://imgur.com/a/qSj0f)).

This installer should both look familiar to Windows users used to the wizard-based installers and be able to be wrapped in a [Chocolatey ](https://chocolatey.org/packages)package dependent on the [JRE](https://chocolatey.org/packages/javaruntime) (if the installer is run with the /S command line switch, it will install silently without requiring GUI interaction).

To build this branch in Jenkins, the `makensis` command needs to be available on the build worker (it works on Linux without Wine too, you can find it in the mayor distribution's repositories). Also, the Launch4J wrapper is now built during the build process by a Gradle task instead of being checked into Git as a binary; this however should not require additional packages on the build worker since the Gradle task automatically downloads the required binutils (*ld* and *windres*) to a local temp folder. Also, the launcher's icon (which the Gradle task puts on the generated exe) is now the Gooey icon instead of the old one; if you prefer the old one, it can be simply replaced by changing the `win-setup/resources/images/icon.ico` file.

Checklist of the work I made on this branch:
- [x] More information in Add/Remove programs (website, license, etc)
- [x] Better graphics (Terasology-related image in the installer pages)
- [x] Icon
- [x] In the uninstaller, allow user to remove game data
- [x] The installer is localizzabile with .properties files; English and Italian strings are included in this PR.
- [x] [Fix the auto updates](https://github.com/MovingBlocks/Terasology/issues/2865#issuecomment-292726759) - there is now a polymorphic update system, which has (at the moment) two implementations, one for the default self-updater and one which downloads the new installer and runs it. It can be further extended for different packaging methods (see [forum post](http://forum.terasology.org/threads/terasology-packaging-linux-distros-windows-mac-discussion.1847/#post-14963)). The update strategy to use is chosen according to an `updateStrategy` file which the packaging systems (the NSIS installer) put in the launcher installation directory; when updating, the launcher reads this file and decides how the update has to be performed. If no `updateStrategy` file is found, the default action is to use the built-in self-updater.
- [x] ~~Setting up the Chocolatey package~~ will be done in separate branch

## Before merging
The NSIS installer is built as part of the `createRelease` Gradle task. Jenkins build server(s) which will execute the builds for release need to have the `makensis` program installed (and on the system path), to allow the build process to compile the script to an executable installer. The script is set up to build [Unicode installers](http://nsis.sourceforge.net/Docs/Chapter1.html#intro-unicode), which support more languages without character display issues; this feature, however, has been introduced in NSIS 3.0 while the NSIS version shipped with some Linux distributions (such as Ubuntu) is still 2.x; if the Jenkins builder runs on a distribution without NSIS 3, it needs to be [built from source](http://nsis.sourceforge.net/Docs/AppendixG.html), or otherwise we can remove the Unicode feature. On Arch Linux, [nsis 3.01 is in the AUR](https://aur.archlinux.org/packages/nsis/), I tested it and it takes a while to compile but works flawlessly.

GooeyHub commented:
Can one of the admins please verify this patch?

skaldarnar commented:
@GooeyHub ok to test

@gianluca-nitti The Chocolatey package can be done in a separate PR (or probably _should_ be in a separate PR). Currently we are creating an .exe with Launch4J - does this become obsolete with the new installer? 
Another question: what happens if the user uninstalls the launcher? Is it possible to ask the user whether he wants to delete the application data (like downloaded games) as well?

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

gianluca-nitti commented:
> Currently we are creating an .exe with Launch4J - does this become obsolete with the new installer?

No, it doesn't - the installer detects the presence of a JRE at install time, while the Launch4J launcher checks for it at launch time. In my opinion it's good for Windows users to have a .exe to launch instead of a .jar since it can have an icon and should work also if there is no file association entry for .jar files. Also if the user uninstalls the JRE after having installed the application, Launch4J shows meaningful information asking to install an appropriate version of Java when trying to launch.
The JRE/JDK check I put in the installer isn't really necessary; without it the user would be able to install the application without having Java installed, but when trying to run, Launch4J would obviously complain about the missing dependency. I made it for the following reasons:
* I think it's better to know that you can't run a program *before* installing it rather than *after* having already installed it
* If I recall correctly Launch4J doesn't warn the user if it's starting a 32-bit JVM on a 64-bit system, while [the checker in the installer does](http://imgur.com/qvVEFZs).

The links created by the installer (in Start Menu and on desktop; both can be disabled during the installation) link to the Launch4J exe, so I think it should be kept.

> Another question: what happens if the user uninstalls the launcher? Is it possible to ask the user whether he wants to delete the application data (like downloaded games) as well?

Tricky point here. The current implementation only deletes the launcher (installed by the installer). I've been thinking about it for some days but I haven't come to a clean solution yet, here are some considerations:
* The launcher allows the user to choose two directories, which will be used, respectively, for game assets and for game data (saves, config, etc). The paths to these directories are stored (on Windows) in `%appdata%\TerasologyLauncher\TerasologyLauncherSettings.properties`.
* Thus, what the uninstaller should do (if the user chooses to delete everything):
    * Read `%appdata%\TerasologyLauncher\TerasologyLauncherSettings.properties` to get the paths to the directories
    * Delete the directories (the user could choose if they should both be deleted, or if only data directory should be kept, etc)
    * Delete `%appdata%\TerasologyLauncher`
* To accomplish this, it would be necessary to:
    * (a) Implement a parser of the properties file in the NSIS language (unhandy but feasible) or alternatively (b) add a command line switch to the launcher (e.g. `--printDirectories`) which reads the files, prints the paths to stdout and then quits without opening the GUI (the NSIS uninstaller would then call it to get the paths to delete)
    * Know where `%appdata%\TerasologyLauncher` exactly is - this is the real problem, because multiple users could use the launcher, which is installed globally, and thus multiple properties file would be present for multiple users. If (b) or a basic implementation of (a) which uses the API calls to get the location of `%appdata%` is chosen at the previous point, the directories would be deleted only for the user which is running the uninstaller. In my understanding, deleting the data for all users would only be possible by enumerating the users and "guessing" where their `%appdata%` is (if I recall correctly since Vista it's always in `C:\Users\<username>\AppData\Roaming`, but it's not guaranteed), definitely not a clean solution in my opinion. This is a [quite old and common problem](https://social.msdn.microsoft.com/Forums/vstudio/en-US/2e072949-10f8-43ef-b557-1042cac72597/how-to-get-localappdata-directory-path-for-each-user-on-perticular-machine?forum=vcgeneral). Someone at Microsoft suggests [not to delete other user's data](https://blogs.msdn.microsoft.com/oldnewthing/20070917-00/?p=25103), however in our case this "data" won't simply be a handful of configuration files but the entire game assets.

Possible solutions:
* Just inform the user that only the launcher is going to be uninstalled and that game and data directories needs to be deleted manually.
* Just remove (after asking permission) the data for the user running the uninstaller as mentioned in Raymond Chen's article I linked before:
> Now, if you want, you can clean up the per-user data for the current user (after asking for permission), but you definitely should not be messing with the profiles of other users.
* Refactor the installer to install the launcher locally (per-user) too, such as in `%appdata%` rather than in `%programfiles%`. The [Squirrel project](https://github.com/Squirrel/Squirrel.Windows) does something like this and is used for example to install the Atom editor. Should be quite easy to do with NSIS too, I'd have to change some of the `SetShellVarContext` lines and the location of some registry keys (which would go in HKCU instead of HKLM). I'm not sure, however, how well this would play with Chocolatey as the registry of installed packages is per-machine and not per-user; could be solved by making a separate package for Chocolatey based on the zip and not on the installer, but then we would be back to the same problem - what data directories should be deleted by the package uninstall script?

skaldarnar commented:
Thanks for the clarification on the launch4j exe, we'll keep it then! 

Another option I see for solving the second issue is to give the launcher itself a way to clean installation data. There should be both an option in the launcher and a command line flag. The uninstaller could simply run `TerasologyLauncher --removeData` after asking for user permission to delete the data of the current user. That way the uninstaller would not need to parse the configuration files.

gianluca-nitti commented:
> There should be both an option in the launcher and a command line flag. The uninstaller could simply run` TerasologyLauncher --removeData` after asking for user permission to delete the data of the current user. That way the uninstaller would not need to parse the configuration files.

Okay, I think I can implement it (will take some days though, I used some basic JavaFX before but I have to revise it a bit to implement the option in the launcher itself). Unfortunately, this still doesn't really solve the problem of the data not being deleted for the other users during uninstall; I'll make it clear in the uninstaller UI that before uninstalling, if you want to delete all the data put on the machine by the launcher and there are other users which used it, you first have to login as every user, open the launcher and use the "Delete data" button that I'll add.

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

skaldarnar commented on src/main/java/org/terasology/launcher/gui/javafx/CleanupController.java:
```diff
@@ -0,0 +1,88 @@
+    @FXML
+    public void cancelCleanupAction(ActionEvent event) {
+        //TODO: is this the right way to close?
+        ((Node) event.getSource()).getScene().getWindow().hide();
```
Seems to be the right way to do it, see: http://stackoverflow.com/questions/13567019/close-fxml-window-by-code-javafx

gianluca-nitti replied:
Okay, thanks for the clarification; I simply copied that code from the method used to close the Settings windows and I had some doubts since the name and the [documentation](https://docs.oracle.com/javase/8/javafx/api/javafx/stage/Window.html#hide--) seems quite misleading to me, since every time a window is opened [we create a new instance of the stage](https://github.com/MovingBlocks/TerasologyLauncher/pull/401/files/d416da3d71db8062cc68e2065379b81705b47e9b..55d90680a8cf1dcebd79832e51fb70c0e216c07c#diff-a9056608efd45f70476a138151ba8260R278), and then to close we "just" hide it. However, from [this answer](http://stackoverflow.com/a/27895501) seems that hiding actually allows it to be garbage collected.

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

gianluca-nitti commented:
I hit an obstacle: to perform the cleanup (delete launcher config, game and data directories), the uninstaller needs to execute the launcher with the `--cleanup` command line argument that I implemented; however, it needs to **wait** for this process to exit before uninstalling (it should also check the return value to see if the cleanup failed and ask the user if he wants to uninstall anyway in this case). The problem is that the current Launch4J wrapper returns immediately (I'd say it "forks" to start the JVM, but we are on Windows so it's not the right term), so the parent process can't really wait for the cleanup to finish.

It should be possible to solve the problem by changing the Launch4J configuration: the `<stayAlive>` entry in the Launch4J configuration can be used to specify that the wrapper must wait for the JVM to exit before exiting itself and return the JVM's exit code, which is exactly what I need ([source](http://launch4j.sourceforge.net/docs.html#Configuration_file) - scroll down to "`<stayAlive>`").

However, at the moment the Launch4J wrapper is [checked in as binary blobs](https://github.com/MovingBlocks/TerasologyLauncher/tree/develop/src/dist) in the repository instead of being built by the Gradle script; to accomplish this I need the Launch4J xml configuration file used to build those binaries. I searched [the whole MovingBlocks organization](https://github.com/search?q=org%3AMovingBlocks+launch4jConfig&type=Code) but I couldn't find the file for this launcher, only one for DestinationSol and one for TeraEd. I found [this repository](https://github.com/MovingBlocks/TerasologyJavaLauncher) which could be related, however I'm not sure what exactly is there - it seems a custom warpper, however @skaldarnar said the launcher is currently using Launch4J.

So I'm kindly asking if anyone has this configuration file or if I should rewrite it from scratch (and, if this is the case, if there is anything I should take in consideration beside setting the title, icon and path to jar file) and if it's appropriate that I remove the binaries from the repository and instead insert the Launch4J configuration script and instruct Gradle to generate the wrapper at every build.

Cervator commented:
Pinging @shartte to see if he's still around - I was uncertain if he only worked on the Terasology exes but it does look like he checked in the most recent Launcher exes too: https://github.com/MovingBlocks/TerasologyLauncher/commit/8c1ea42c80bb83cd332f927ffbae19a61fa0fcbb

So maybe he has or remembers something about the Launch4J config. We still have some partial work (that other repo) aimed at executables that better embed a JRE and/or detect a proper Java 8. That's been quiet for a long time though. https://github.com/MovingBlocks/DestinationSol also has its own executables that *do* at present work with a fully embedded JRE, might be worth looking at.

I'd be happy to see the exes replaced by the config files + a Gradle task to build them as needed instead :-)

shartte commented:
Hey @Cervator, I think the landscape has somewhat changed.
Since there are now "good" OpenJRE builds for Windows, one might simply go and dissect those for inclusion in Terasology.
The other alternative is waiting for Java 9, and going straight ahead and embed that (sans any modules that are not needed for Terasology).
But boy the Launch4 configs... I have to go and look for those. I *think* I have them checked in somewhere.

shartte commented:
Okay after checking (this was two years ago heh), are you certain the launchers are actually still Launch4j?
I think I actually rewrote them in C back then:
https://github.com/MovingBlocks/TerasologyJavaLauncher

skaldarnar commented:
> however, it needs to wait for this process to exit before uninstalling (it should also check the return value to see if the cleanup failed and ask the user if he wants to uninstall anyway in this case). The problem is that the current Launch4J wrapper returns immediately (I'd say it "forks" to start the JVM, but we are on Windows so it's not the right term), so the parent process can't really wait for the cleanup to finish.

Could you just start the launcher via Java directly (not using the .exe at all) for the uninstall process? 

@shartte @gianluca-nitti I'm not sure we are (still) using launch4j-  there's a good chance your re-writes made into the repository. In that case, can you give a hint what needs to be changed in order to keep the .exe process alive?

gianluca-nitti commented:
Thank you all for the interest. Some updates:
* @shartte I cloned your repository and had a look at the binaries in `dist/`; the `TerasologyLauncher.x86.exe` and `TerasologyLauncher.x64.exe` there and in this repository's `src/dist/`, which I presume are built from the C/C++ sources in your repository, are indeed identical (checked SHA1s).
However, seems like at the moment they are ignored by the Gradle build (see [this line](https://github.com/MovingBlocks/TerasologyLauncher/blob/develop/build.gradle#L373)); in fact, the ZIP file generated when executing the `createRelease` task only contains `TerasologyLauncher.exe`, taken from `src/dist/`. This one really looks like it's Launch4J - [here](https://gist.github.com/gianluca-nitti/bd5ea3052fe38dfbd2c14eb5c7480fbe#file-output-os-strings-on-terasologylauncher-exe-L52) is the output from `strings` (Mingw binutils), see [lines 52 and next ones](https://gist.github.com/gianluca-nitti/bd5ea3052fe38dfbd2c14eb5c7480fbe#file-output-of-strings-on-terasologylauncher-exe-L52).
* @skaldarnar Yes I thought about spawning a JVM directly from the uninstaller, however I can't be sure the user has `java`/`javaw` on the PATH and I'd have to look in the registry keys to have a safe way of knowing where the Java home is; if I recall correctly this is exactly what Launch4J does, so I didn't want to duplicate it, but I could go this route anyway if you prefer.
* @Cervator I thought we were moving away from the embedded JRE (see [here](http://forum.terasology.org/threads/terasology-packaging-linux-distros-windows-mac-discussion.1847/#post-14942) and [here](https://github.com/MovingBlocks/Terasology/issues/2865#issuecomment-292708982); the code in this PR already implements a check for the JRE before installing and provides instructions to download it if it's not installed or if it's outdated) but if you still prefer to bundle it, we could discuss how to change the build process and the installer to do it.
> I'd be happy to see the exes replaced by the config files + a Gradle task to build them as needed instead :-)

Ok, I could redo the Launch4J config from scratch and set up Gradle to build the exes if everyone is OK with this (I'd have to remove all the exes in `src/dist/`); should be quite simple since [there is a plugin](https://plugins.gradle.org/plugin/edu.sc.seis.launch4j), however would probably require to install more packages on the Jenkins builder to work (assuming it's Linux, it would require a minimal cross-toolchain for windows with at least `ld` and `windres`, see [here at the bottom of the page](http://launch4j.sourceforge.net/)).

Funnily enough, [seems like I can disable the stayAlive feature](http://launch4j.sourceforge.net/docs.html#Runtime_options) via a simple command line switch at runtime (`--l4j-dont-wait`) but not enable it without rebuilding the wrapper.

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

skaldarnar commented:
I would go for:
* use launch4j and integrate it with gradle (plugin) to generate .exe
* don't bundle the JRE but manage it as dependency/check during install
* use the new .exe in the uninstall process

IMHO this is a clean solution which should work for most users.

gianluca-nitti commented:
@skaldarnar perfect, that's exactly what I made in the [latest commit](https://github.com/MovingBlocks/TerasologyLauncher/pull/401/commits/facd9278954c7c6b9978d0d5c0ad90455b888601). I used [this handy Gradle plugin](https://github.com/TheBoegl/gradle-launch4j) and the Launch4J configuration (few lines) is actually embedded in the gradle script (the `createExe` task).

Cervator commented:
Heya @shartte good to hear from you :-)

@gianluca-nitti on bundling the JRE - I think it is appropriate in some future cases like if we get Terasology onto Steam like Destination Sol. In that case as part of packaging a Steam release you have to include the JRE as a build depot thing with executables configured to use it.

That may not quite be the same thing as we've talked about in the past though. In short I am unsure and am happy if we're covering all scenarios via other means like package managers :-) Users come in all shapes and sizes though, some will be dead set against using a game launcher or maybe even a game installer. Somebody might want a zip they can put on a fresh system and just run the game. Hard to please everybody.

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

gianluca-nitti commented:
As you can read in the edited first post, this is no longer work in progress; I added everything I had in mind.

If this gets merged, I'll develop a Chocolatey package in a separate PR.

Any comment is welcome, it's the first time I make a pull request this big (> 1000 lines) and I apologize if there is anything wrong (code style, etc).

Testing the new update system was quite challenging since there obviously is no .exe package on the published releases on the real Jenkins server, so I had to set up a simple server (I used Node.js) to mock the Jenkins behavior for certain URLs, run it locally and add `jenkins.terasology.org` to my hosts file to redirect the launcher's requests to the local server. I added a "fake" versionInfo.properties file to the installed jar to simulate the availability of a new version and verify that the update process works correctly. [Here is the simple server](https://gist.github.com/gianluca-nitti/72ac67c8697c9fbdda8901495a4f7248) if you are interested; it needs the installer which will be sent as the update in the working directory too. I'm sorry if this is a "noobish" and not quickly reproducible way to test it, but I don't know how a proper automated integration test should be developed for this (I'm not even sure it's possible since it would require to simulate the click on the UAC dialog).

gianluca-nitti commented:
Little update: I added the instructions about the correct NSIS versions required for the build [in the main PR description](https://github.com/MovingBlocks/TerasologyLauncher/pull/401#issue-220389615).

skaldarnar commented on build.gradle:
```diff
@@ -372,8 +399,42 @@ task copyHtml(type: Copy) {
+    into tmpDir
+}
+
+task renderNSISScript(type: Copy) {
```
is "render" the correct terminology? I would go for _run_ or _execute_.

gianluca-nitti replied:
I named it "render" because this task doesn't actually compile the script (this is done by `distNSIS`, which depends on this one); `win-setup/win-setup.nsi` is a *template* and this tasks transforms it into an intermediate product, a valid NSIS script which is placed at `${buildDir}/tmp/distNSIS`, by "filling the holes" with the right parameters (language strings, version information, etc - see lines below), which is then compiled by the `distNSIS` task. This is accomplished using the [expand method of the copy task](https://docs.gradle.org/3.4.1/dsl/org.gradle.api.tasks.Copy.html).

So I used "render" which is commonly used with templates, but I can change it if you want.

skaldarnar replied:
> So I used "render" which is commonly used with templates, but I can change it if you want.

Thanks for the explanation. I believe your words on the template thing here 😉 Leave it as `render`

skaldarnar commented on build.gradle:
```diff
@@ -372,8 +399,42 @@ task copyHtml(type: Copy) {
+    dependsOn renderNSISScript
+    dependsOn copyNSISresources
+
+    executable "makensis"
```
what happens if `makensis` is not installed? Can we fail gracefully in that case with a good error message?

gianluca-nitti replied:
Actually fails with the following message:
```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':distNSIS'.
> A problem occurred starting process 'command 'makensis''

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

```

Not sure if it's graceful enough, if it isn't please provide me with the message you'd like to see so I can improve it.

skaldarnar replied:
hmm. I think this is obvious enough (for now). We can always improve on little things, but one gets the information that there is a command which could not be found. Should be enough to solve the problem.

skaldarnar commented on src/main/java/org/terasology/launcher/TerasologyLauncher.java:
```diff
@@ -96,33 +102,91 @@ public void start(final Stage initialStage) {
-                    }
-                } catch (IOException | LauncherStartFailedException e) {
-                    openCrashReporterAndExit(e);
+        launcherInitTask.setOnSucceeded((final WorkerStateEvent workerStateEvent) -> {
```
I like that change! :+1:

skaldarnar commented on src/main/java/org/terasology/launcher/TerasologyLauncher.java:
```diff
@@ -96,33 +102,91 @@ public void start(final Stage initialStage) {
     }
 
+    private void cleanupStart(Stage stage) {
+        logger.info("Starting cleanup tool because --cleanup argument was specified");
```
Maybe just print "Starting in clean-up mode" to be more generic here. If you want to log the command line argument, do that when checking the parameters.

skaldarnar commented on src/main/java/org/terasology/launcher/gui/javafx/CleanupController.java:
```diff
@@ -0,0 +1,119 @@
+                return true;
+            } catch (IOException e) {
+                logger.warn("Failed to delete {}", f.getAbsolutePath());
+                retry = GuiUtils.showBinaryChoicheDialog(BundleUtils.getMessage("cleanup_error"),
```
typo: `showBinaryChoicheDialog` -> Choice

skaldarnar commented on src/main/java/org/terasology/launcher/gui/javafx/CleanupController.java:
```diff
@@ -0,0 +1,119 @@
+        cancelButton.setText(BundleUtils.getLabel("cleanup_cancel"));
+    }
+
+    private boolean tryDeleteDirectory(CleanupUtils.Directory directory) {
```
just call it `deleteDirectory`

skaldarnar commented on src/main/java/org/terasology/launcher/updater/AbstractLauncherUpdater.java:
```diff
@@ -55,7 +50,11 @@
 
-    public LauncherUpdater(TerasologyLauncherVersionInfo currentVersionInfo) {
+
+    // constructor and getters are for use by subclasses (default access level -> access allowed only in the updater package)
```
can remove this comment.

skaldarnar commented on src/main/java/org/terasology/launcher/updater/LauncherUpdaterFactory.java:
```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2016 MovingBlocks
```
oh, we have 2017 already :stuck_out_tongue_closed_eyes:

gianluca-nitti replied:
Ah, right :)

Should I update it on all the files I worked on, or only on the new ones created on this branch?

skaldarnar replied:
you can update them all...

skaldarnar commented on src/main/java/org/terasology/launcher/updater/LauncherUpdaterFactory.java:
```diff
@@ -0,0 +1,83 @@
+
+    /**
+     * @return The right updater for the running instance of the launcher.
+     * It is detected by reading the updateStrategy file in the launcher installation directory.
```
Should we keep _updateStrategy_ as separate file or put it as option in the launcher settings somehow? Would it make sense to change it after packaging? 
(I guess the current solution is fine, just wanted to ask)

skaldarnar commented on src/main/java/org/terasology/launcher/util/GuiUtils.java:
```diff
@@ -68,6 +71,18 @@ public static void showInfoMessageDialog(Stage owner, String message) {
         showMessageDialog(Alert.AlertType.INFORMATION, BundleUtils.getLabel("message_information_title"), message, owner);
     }
 
+    public static boolean showBinaryChoicheDialog(String title, String message, String btnTrue, String btnFalse) {
```
typo -> Choice

skaldarnar commented on src/main/resources/org/terasology/launcher/bundle/LabelsBundle_en.properties:
```diff
@@ -126,6 +126,8 @@ settings_launcher_chooseLanguage=Choose language
 settings_launcher_closeLauncherAfterGameStart=Close launcher after game start
 settings_launcher_downloadDirectory=Download directory
 settings_launcher_downloadDirectory_open=Open
+settings_launcher_cleanupDirectories=Delete directories created by the launcher
```
I would just call it "Cleanup" or something as short

skaldarnar commented on src/main/resources/org/terasology/launcher/bundle/LabelsBundle_en.properties:
```diff
@@ -126,6 +126,8 @@ settings_launcher_chooseLanguage=Choose language
 settings_launcher_downloadDirectory=Download directory
 settings_launcher_downloadDirectory_open=Open
+settings_launcher_cleanupDirectories=Delete directories created by the launcher
+settings_launcher_cleanupDirectories_open=Delete various directories created by the launcher (cleanup)...
```
This is a pretty long button text… I would go for "Delete Launcher Data"

gianluca-nitti commented:
@skaldarnar Thank you very much for the in-depth review!

For all the comments I've not replied directly - will soon make a commit with the requested changes.

qwc commented:
Nice work @gianluca-nitti !

I approve! :+1: 

@skaldarnar Of course we can provide a build node with the appropriate packages installed, in this case this would be a archlinux docker image I suppose. I would fit good to the linux game packaging I am already putting into docker images for easier portability to other CI environments.

Only question I have is, that when this is a gradle task which has to be called from createRelease, is it even possible to put that task to another buildnode?
I would put it into an own target/task, which can be called independently, so we can just go with an docker image containing nsis and maybe gradle (java too?) if needed. Copy all artifacts and just call the distNSIS task.
(I may be too long out of gradle experience to know exactly how this could work... should refresh my knowledge sometime near...)

gianluca-nitti commented:
> I would put it into an own target/task, which can be called independently, so we can just go with an docker image containing nsis and maybe gradle (java too?) if needed. Copy all artifacts and just call the distNSIS task.

This should already be feasible - this is the definition of `createRelease`:
```
task createRelease() {
    dependsOn distZip
    dependsOn distNSIS

    doLast {
        println 'Created release: ' + displayVersion
    }
}
```

So `./gradlew createRelease` is basically a shorthand for `./gradlew distZip && ./gradlew distNSIS`. If you want to exclude NSIS, the main build node could simply call `distZip` instead of `createRelease`, and the NSIS node just `distNSIS`; if the artifacts are already there (i.e. the `build/` directory is copied form the main build node to the NSIS build node), Gradle should not repeat the tasks `compileJava`, `distZip`, etc because their targets are already up-to-date.

It sounds somewhat unusual to me copying the artifacts from a build node to another one but I don't see anything wrong with it.

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

gianluca-nitti commented on build.gradle:
```diff
@@ -146,6 +153,24 @@ def convertGitBranch = { gitBranch ->
+    version = "${versionMajor}.${versionMinor}"
+    textVersion = displayVersion
+    companyName = orgName
+    jreMinVersion = "${targetCompatibility}.0" // launch4j requires it's in the form x.x.x, thus this currently renders to 1.8.0
```
Uhm, I'm not sure how to fix this - JRE/JDK versions are in the form x.x.x_[xx] (where each x is a single digit), so if I try with `1.8.40` the build fails with `launch4j: Invalid data: Minimum JRE version should be x.x.x[_xx]`, what version should I exactly require? `1.8.0_40`? I'm sorry but I can't find any `build.sbt` file, or resources showing that a specific version is required for JavaFX.

Also if a specific update of Java 8 is required I'll have to change the JVM detection logic in the installer, because the `CurrentVersion` value in the registry key `HKLM\SOFTWARE\JavaSoft\Java Runtime Environment` only specifies `1.8`, while I have to look in the subkeys to find the update number (e.g. 121 on my machine which has JRE 1.8.0_121).

skaldarnar replied:
I'm sorry for `build.sbt`, got confused with our work build system here - I totally meant the `build.gradle`. In addition, you are right with `1.8.0_40` (as you can see [here](https://github.com/MovingBlocks/TerasologyLauncher/blob/develop/build.gradle#L53)). 

Please excuse this little inconvenience from my side ☺️

skaldarnar replied:
Oh, and it may be because of Dialogs: http://code.makery.ch/blog/javafx-dialogs-official/ (not 100% sure, but there was a reason for this requirement).

qwc commented:
> It sounds somewhat unusual to me copying the artifacts from a build node to another one but I don't see anything wrong with it.

Well that's common when you do not want that the jenkins master or specific nodes have all dependencies installed to build every piece of the project. It's often easier to provide VMs/docker-images for different stages of the project build setup. Just like I've setup two docker images only for packaging.
With a java project it's easy to compile the source for all platforms on one build slave, now just think of a platform independent C++ project, It has to be built on all platforms it should support to deliver the binary, some may be done through crosscompiling, but not all. VMs/docker for different tasks...

gianluca-nitti commented:
Ok, I agree.

To recap, these are the changes I may need to do on this branch:
- [ ] @qwc: let me know if The current set of tasks is appropriate (in this case, to split the build and packaging between build nodes, the main one needs to be configured to call `distZip` instead of `createRelease`) or if you prefer   to have `createRelease` to depend only on `distZip`.
- [ ] @skaldarnar: need clarification on the min JRE version (see [here](https://github.com/MovingBlocks/TerasologyLauncher/pull/401#discussion_r112897054)) - more specifically, I need to know if by 1.8.40 you meant 1.8.0_40 and if you can link some source stating that JavaFX requires that version (I did some research but without success).
- [ ] (this may be a silly question, but I'm not sure if there is some standard in the project so I ask) whether I have to update the copyright year in the license snippet at the top of the source files only for the files *created* in this patch or for all the files that have been *modified*.

skaldarnar commented:
@qwc I'm fine with an ArchLinux node, can you set up the image? I guess just Java and NSIS installed should be enough as the local gradle wrapper should be used (no need to install Gradle system-wide).

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

qwc commented:
@skaldarnar Yep. Will do so over the weekend, I'm kinda busy currently.

- [ ] Setup ArchLinux docker build node for NSIS Build (with java and nsis 3.x)

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

gianluca-nitti commented:
I fixed the version checker (now it requires 1.8.0_40 instead of just 1.8) both for Launch4J and the installer.

Since working with version strings in NSIS is quite unhandy, to make the installer perform this check I made a simple Java program to check the version of the JVM it's running on ([repo here](https://github.com/gianluca-nitti/jvm-version-check)). It takes a version on the command line and returns 0 or 1 exit code depending on the actual JVM version being >= the given one or not. The installer includes this program (a ~2k jar), and extracts it as a temporary file (which is then deleted both on success and on failure) and executes it with each JVM found on the system, and determines if a suitable JVM is found by looking at it's exit code.

I included the jar in the repository; I know this may not be the best option, but I went this way because I'm not sure if there is a real alternative which makes sense:
- Including the sources - but then who should to compile it? Main build node or NSIS build node? Should I make an entire Gradle subproject just to compile this single class?
- Pushing it to a binary repository (like [JitPack](https://jitpack.io/)) and get it via Gradle - feels strange, the modules Gradle downloads are Gradle plugins or libraries that must go to the project classpath

So I put it in the repository since after all this is just a tiny jar and semantically is a resource for the installer exactly as the icon/splash image are. If you don't like it and know a better way to include it, I'll edit accordingly.

Cervator commented:
So I'm finally back home and more able to start catching up with these kinds of things again. I'm sure I'm missing a bunch of stuff so feel free to poke me elsewhere if my attention is needed, or even better link me to things on Slack :D

I was playing with Jenkins Pipeline for work while I was away, so lots of new ideas on how to structure stuff, we're even using some Chocolatey there, as well as Puppet to prepare build nodes. Old school you'd make a separate Jenkins job for the NSIS part that's forced onto a specific node and then trigger it from the main job, new school you can write in node requirements and parallel execution really nicely in a single pipeline job. 

One question that remains for me is whether to split Gradle tasks between nodes if they can be run in parallel, as opposed to rely on Gradle itself. I know you can enable a parallel flag for unit testing to split them out into multiple threads, and you can probably find a way to indicate tasks can be run in parallel, like findbugs + Checkstyle, but that doesn't get them onto different nodes. Which may be overkill anyway, especially if you then get the added overhead of analyzing the entire project tree again before you can begin executing. Probably we can worry about that if we later need to make builds faster.

With a Jenkins pipeline in particular you can more easily share the load (and files!) across nodes. There is a new "stash" command you can use on a fileset early in the process you can then unstash later on a different node. If we need that for the NSIS piece then that might be useful - or just build the launcher on there entirely in the first place?

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

skaldarnar commented:
@gianluca-nitti Excuse my long absence. I'm not bothered by the JAR you've added. I would just ask you to put a note in the README and/or [wiki](https://github.com/MovingBlocks/TerasologyLauncher/wiki/Documentation) and link to the repository the JAR is compiled from. That way, we know where it came from, how we can recompile it in the future, and where we have to look should we get into a situation in which we need to get our hands on it. 

@qwc @Cervator I'm fine with merging this whenever the devops perpective looks alright. What work is still open in order to get this properly set up in the build pipeline? Merging it just now should not break anything (if I see it correctly), but we would not see any difference, either.

gianluca-nitti commented:
Hi there, no problem for the "pause" on this issue.

Good idea to put a link to the [repository](https://github.com/gianluca-nitti/jvm-version-check) the additional jar is built from. I'm not sure where it should go, however; I couldn't find any section in the main README and in the wiki related to libraries/dependencies, the closest one could be the [License](https://github.com/MovingBlocks/TerasologyLauncher#license) section, but I'm not sure if that's the right place. Alternatively I could put a small secondary README file right in the [directory where the jar is](https://github.com/gianluca-nitti/TerasologyLauncher/tree/NSISinstaller/win-setup/resources).

> Merging it just now should not break anything (if I see it correctly), but we would not see any difference, either.

Giving a small feedback even if this wasn't directed to me, I fear it would break the build as of now - this branch is passing because (correct me if I'm wrong) it's being built by the [TerasologyLauncherPRs](http://jenkins.terasology.org/view/Launcher/job/TerasologyLauncherPRs/lastBuild/console) job, which "only" runs to the `check` target, while [the job which builds the develop branch](http://jenkins.terasology.org/view/Launcher/job/TerasologyLauncher/lastBuild/console) also runs `createRelease`, which after merge will depend on `distNSIS` too, most likely causing the build to fail if `makensis` is not on the system path.

I made `createRelease` dependent on `distNSIS` because originally I thought that NSIS would have been installed on the main worker node(s). Given that `distNSIS` works by extracting the zip produced by `distZip` if we want to keep the nodes separate, things should work ok if I remove this dependency (thus leaving `distNSIS` a task that must be invoked externally, i.e. no tasks depend on it) and there is a stash-like functionality across nodes as said by @Cervator in the previous post.
The flow would be something like: build and make zip (`createRelease`, which wouldn't depend on `distNSIS` anymore) on the main node -> stash the zip -> unstash it on the NSIS node -> build the installer (`distNSIS` on the NSIS node).

Looks a bit overly complicated to me but should be feasible.

qwc commented:
IMO, it would be nice to remove that depdendency so that the standard build can be done without hassle or changes for the build node. And add that build node for the nsis build later on.

I'm quite behind my own proposed schedule for the packaging, but as always plans are interrupted by life. :)
When I have time for it, I'll add a nsis node for building the installer.

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

gianluca-nitti commented:
Okay thanks for the input, now `createRelease` doesn't automatically try to build the NSIS installer too; to do so, the `distNSIS` task must be invoked separately, and merging this *shouldn't* break the build.

Also added a README in the [win-setup/resources/](https://github.com/gianluca-nitti/TerasologyLauncher/tree/68241f9b6d3de98ebe50f72435d29a2486bfeb2b/win-setup/resources) folder with the link to the sources for jvm-version-check.
